### ISSUE TITLE
Input value can't be simply controlled by `value` prop

### ISSUE THREAD

**Issue by yqrashawn** [zeit-ui/react#335]
# Bug report 🐞
Not sure if this is a bug or a design decision. In the scanbox link, keep input in either input, you can see the <input>'s value is controlled by the `value` state while the <Input> is not.

## Version & Environment

  - Version of browser chrome Version 83.0.4103.116 (Official Build) (64-bit)
  - Version of `zeit-ui/react` check sandbox

## Expection

Just want to know if this is intentional.

**Comment by unix** [comment]
I'm not sure what you mean, please provide more information or an online link for this issue.

**Comment by yqrashawn** [comment]
Sorry for that, forgot to add the sandbox link. Here it is.

https://codesandbox.io/s/zen-colden-byknw?file=/src/App.js

**Comment by unix** [comment]
Because `Input` component is a controlled component, refer to:

> With a controlled component, the input’s value is always driven by the React state. While this means you have to type a bit more code, you can now pass the value to other UI elements too, or reset it from other event handlers.
  - [Controlled component](https://reactjs.org/docs/forms.html#controlled-components)
  - [Uncontrolled component](https://reactjs.org/docs/uncontrolled-components.html)

**Comment by GuichiZhao** [comment]
<img width="885" alt="截屏2020-07-16 上午10 56 36" src="https://user-images.githubusercontent.com/12082547/87653375-0618f500-c788-11ea-81aa-5fa04e935710.png">
Tell me the value of  `value` when the input bar displaying '456aa'

**Comment by unix** [comment]
@GuichiZhao 
It's very nice to point out the issue.

This is because every time the `value` set is exactly the same, `useEffect` ignores the tracking of changes. If you `setValue` different in the `handler`, everything is OK.

This is rare in other component libs because they use the `class` component. If `setValue` is used in the `class` component, the setting action of value can be captured no matter whether it changes or not. There is no particularly good solution to this problem when we use a controlled component (**If you have to set the same value**), I created a pull request to solve this issue, I try to ignore all changes of `value` to capture each set action, what do you have about this?

**Comment by unix** [comment]
Now, before the issue is solved, as an alternative, you can always synchronize `value` or choose to use `useInput` to synchronize automatically:

1. Synchronously manually.
```jsx
const handler = (e) => {
    setValue(e.target.value)
}
```

2. Synchronously with `useInput`.

```jsx
const { state, bindings } = useInput('The Evil Rabbit')
useEffect(() => {}, [state])

return <Input {...bindings} />
```

**Comment by GuichiZhao** [comment]
@unix 
If you believed the component is controlled, what should happen when user clicks the `clear` button.

**Comment by unix** [comment]
If you want using *clear*, you have to choose one:
  - `value` has been set: synchronize the `value` of `Input` manually.
  - `value` is not used: nothing needs to be done.

**Comment by GuichiZhao** [comment]
In my option, the controlled way of designing the component can be phased out completely.
Here is a dilemma:
If you choose the controlled way, there is not much the library can provide, the state should be maintained by the client code externally. 
It is good the library can provide some useful functionality such as `clearable`.  In a broader setting, the state can be rather complicated, and user should not care too much about details. Here is the point, the library can only come into play when the component is **not controlled**

However, if the component is not controlled, how it possible the caller needs to read out the states maintained inside the component by the library. It is just a matter of API design, A simple design is to expose an `onChange` API and the user can sync data when they see appropriate. The library author only needs to make sure the callback param of `onChange` can reflect the UI state.

By doing this, we lose some power of React`, say we can not express the UI state using `props` in a simple way. However, even though not as beautiful as `props`, we can still design some sort of imperative API to change the state.

In a complex setting, the library can provide much more than the expressiveness of declarative way promoted by React

It OK to abandon the controlled style prefer a more imperative design of the API the provide more useful functionalities to the users

**Comment by unix** [comment]
Very good discussion about the controlled component. In fact, I agree with you on some points.

First of all, I'd like to talk about why I provide (or design) internal state of the component: (I'm not arguing for point, just stating why) In most component libraries or official examples of React, it is a common practice to synchronize state externally, keeping a similar practice is to hope that most users can get a similar experience when using it.

Secondly, the `Input` component also provides some uncontrolled options:

  - If `value` is not set, no external synchronization is required (you mentioned that too). This means that values can be obtained at any time but `state` maintenance is not required.
  - Forwarded native `Ref`. Most of the time, users can skip synchronization and directly use `Ref` API to get values.
  - Provides `initialValue` options. This is useful if you want to skip synchronization and set the initial value.

[Here](https://codesandbox.io/s/zeit-ui-set-input-by-ref-0xgiv?file=/src/app.js) I've created an example for `Ref` of `Input`.

Of course, even if the user can control `Ref`, the user cannot modify the internal `state` in other ways. The only way to do this is to add `value` and then change the `value` prop.

----

To be sure, now we have two problems that can be improved:
  1. Always need to synchronize `state` externally.
  2. `Value` of internal cannot be modified in an imperative API.

### About synchronization of `value`:

Now we have some solutions to reduce the burden of this problem:

  - With `useInput`, you don't need to synchronize manually at all.
  - Do not set `value`. It may be useful in a few scenarios.
  - Do not set `value`, and use `initialValue` instead. (it can't cover all scenes.)
  - Do not set `value`, and use `Ref` to get value. (but the `set` action is invalid)

After pull request merged, we will add the following  ways:

  - Do not set `value`, and use `Ref` to set value. 

### About imperative API:

Thanks to your advice, I optimized the way `Input` component works. Now when the `value` attribute is not set, the `Input` component is uncontrolled, In other words, you can use external ways to update, such as `Ref` or operate DOM API.

And I added an example of the imperative API to the document page of `Input`: #336

**Comment by GuichiZhao** [comment]
>In most component libraries or official examples of React, it is a common practice to synchronize state externally, keeping a similar practice is to hope that most users can get a similar experience when using it.

point me out the official examples of React. In my option, maintain a state both internally and externally tend to break the data flow somehow


>Here I've created an example for Ref of Input
the user cannot modify the internal state in other ways

After reading the code you provide,  the user  "can" control the internal state without passing through the value `props`, refer to `useimperativehandle` for details. Actually, `useimperativehandle` is the "best" solution in my opinion. The user can make use of the functionalities provided library and can set the state from external without breaking the data flow.

React is a library highlighting the declarative way of programming, i.e. UI is controlled by "property" rather than the "method". However, the imperative way, i.e. change state using method can be useful when user do not need to take full control of the component provided by the library 

Hopefully, I express myself clearly, otherwise, I will provide some further examples and my idea of API design

**Comment by unix** [comment]
If you have your own opinion, would you mind providing a new pull request for this?

**Comment by GuichiZhao** [comment]
Forget about my design options for a while, after reading your comment **vary carefully**, I believed you get React wrong in some aspect rather than a communication issue(i.e. I misunderstand you somehow).

Firstly: the data flow
 >it is a common practice to synchronize state externally,

If you want to take control of a component, pass the `props`. By "synchronize", you mean there still a chance the props and the UI can not be synchronized. In the current `input` component. `value` can flush UI but UI can also be modified in other ways, which means UI can not be reflected by `value`, how `value` useful then. At least it can not make sense to me and there is no similar design off the top of your head.

Secondly: imperative API

In current design, you believe:

>Forwarded native Ref. Most of the time, users can skip synchronization and directly use Ref API to get values.
> the user cannot modify the internal state in other ways

I can summarize the above, user can get the state but not set the state of the component

In the new design, you want:
>Do not set value, and use Ref to set value.

All the above make sense to me.

However, you also mention:
>With useInput, you don't need to synchronize manually at all.

From the doc, I see something like this `const { state, setState, reset, bindings } = useInput('The Evil Rabbit')`. `setState` and `reset` tend to be used in a **imperative** way

Actually, such API also makes sense to me. However, I strongly recommend you check out `https://reactjs.org/docs/hooks-reference.html#useimperativehandle`, i.e. how `useimperativehandle` work if you never do it before


your `useInput` implement exactly the same thing as `ref` and `useimperativehandle`, which already provided by React natively

**Comment by unix** [comment]
I'm not sure what exactly do you want, in your opinion many points can not satisfy you, but you're not willing to create a pull request either. 
I don't want to get into such a debate, I will not update this issue until you create a new pull request to describe your requirements.

**Comment by yqrashawn** [comment]
resolved by #336
